File: mcp-v2-typescript-client/src/client/factory.ts
/**
 * MCP v2 Client Factory
 * Factory functions for creating MCP v2 clients with different configurations
 */

import type { MCPClient, MCPClientOptions, TransportType } from '../types/index.js';
import { BaseMCPClient } from './base.js';
import { createTransport } from '../transport/index.js';
import { createContextManager } from '../context/index.js';
import { createLogger } from '../utils/logger.js';
import { createMetrics } from '../utils/metrics.js';

/**
 * Create a new MCP v2 client with the specified options
 */
export function createMCPClient(options: MCPClientOptions): MCPClient {
  // Create logger
  const logger = createLogger(options.logging);
  
  // Create transport
  const transport = createTransport(options.transport, logger);
  
  // Create context manager
  const contextManager = createContextManager(options.context, logger);
  
  // Create metrics collector
  const metrics = options.metrics?.enabled ? createMetrics(options.metrics) : undefined;
  
  // Create and return the client
  return new BaseMCPClient({
    transport,
    contextManager,
    logger,
    metrics,
    options
  });
}

/**
 * Create a simple HTTP JSON-RPC client
 */
export function createHTTPClient(url: string, options?: Partial<MCPClientOptions>): MCPClient {
  return createMCPClient({
    transport: {
      type: 'http',
      url,
      options: options?.transport?.options
    },
    ...options
  });
}

/**
 * Create a WebSocket client for real-time communication
 */
export function createWebSocketClient(url: string, options?: Partial<MCPClientOptions>): MCPClient {
  return createMCPClient({
    transport: {
      type: 'websocket',
      url,
      options: options?.transport?.options
    },
    reconnect: {
      enabled: true,
      maxAttempts: 5,
      backoffMs: 1000,
      exponential: true,
      ...options?.reconnect
    },
    ...options
  });
}

/**
 * Create a gRPC client for high-performance scenarios
 */
export function createGRPCClient(url: string, options?: Partial<MCPClientOptions>): MCPClient {
  return createMCPClient({
    transport: {
      type: 'grpc',
      url,
      options: options?.transport?.options
    },
    ...options
  });
}

/**
 * Create a client with Redis caching enabled
 */
export function createCachedClient(
  transportConfig: { type: TransportType; url: string },
  redisConfig: { host: string; port: number; password?: string; db?: number },
  options?: Partial<MCPClientOptions>
): MCPClient {
  return createMCPClient({
    transport: transportConfig,
    cache: {
      enabled: true,
      ttl: 300000, // 5 minutes default
      redis: redisConfig
    },
    ...options
  });
}

/**
 * Create a client optimized for development with debug features
 */
export function createDevClient(url: string, options?: Partial<MCPClientOptions>): MCPClient {
  return createMCPClient({
    transport: {
      type: 'http',
      url
    },
    logging: {
      level: 'debug',
      transport: 'console',
      format: 'text'
    },
    metrics: {
      enabled: true,
      interval: 5000
    },
    reconnect: {
      enabled: true,
      maxAttempts: 3,
      backoffMs: 1000,
      exponential: false
    },
    ...options
  });
}

/**
 * Create a production-ready client with full monitoring
 */
export function createProductionClient(
  url: string,
  config: {
    redis?: { host: string; port: number; password?: string };
    metrics?: { endpoint: string; interval?: number };
    auth?: { type: 'bearer' | 'api-key'; credentials: Record<string, string> };
  },
  options?: Partial<MCPClientOptions>
): MCPClient {
  return createMCPClient({
    transport: {
      type: 'http',
      url,
      options: {
        timeout: 30000,
        retries: 3,
        compression: true,
        authentication: config.auth,
        tls: {
          enabled: true
        }
      }
    },
    cache: config.redis ? {
      enabled: true,
      ttl: 600000, // 10 minutes
      redis: config.redis
    } : undefined,
    logging: {
      level: 'info',
      transport: 'remote',
      format: 'json'
    },
    metrics: {
      enabled: true,
      endpoint: config.metrics?.endpoint,
      interval: config.metrics?.interval || 30000
    },
    reconnect: {
      enabled: true,
      maxAttempts: 10,
      backoffMs: 1000,
      exponential: true
    },
    ...options
  });
}

/**
 * Create a client cluster for load balancing and high availability
 */
export function createClusterClient(
  endpoints: Array<{ url: string; weight?: number }>,
  options?: Partial<MCPClientOptions>
): MCPClient {
  // For now, return a single client pointing to the first endpoint
  // In a full implementation, this would create a load-balancing client
  const primaryEndpoint = endpoints[0];
  if (!primaryEndpoint) {
    throw new Error('At least one endpoint is required for cluster client');
  }
  
  return createMCPClient({
    transport: {
      type: 'http',
      url: primaryEndpoint.url
    },
    ...options
  });
}

/**
 * Create a client with custom authentication
 */
export function createAuthenticatedClient(
  url: string,
  auth: {
    type: 'bearer' | 'basic' | 'api-key' | 'oauth2';
    credentials: Record<string, string>;
  },
  options?: Partial<MCPClientOptions>
): MCPClient {
  return createMCPClient({
    transport: {
      type: 'http',
      url,
      options: {
        authentication: auth
      }
    },
    ...options
  });
}

/**
 * Create a client with streaming capabilities
 */
export function createStreamingClient(url: string, options?: Partial<MCPClientOptions>): MCPClient {
  return createMCPClient({
    transport: {
      type: 'websocket',
      url,
      options: {
        compression: true
      }
    },
    reconnect: {
      enabled: true,
      maxAttempts: -1, // Infinite retries for streaming
      backoffMs: 2000,
      exponential: true
    },
    ...options
  });
}

File: packages/mcp-v2-python-client/src/mcp_v2_client/factory.py
"""
Factory for creating MCP v2 clients with different configurations.
"""

from typing import Dict, Any, Optional
from uuid import uuid4

from .client import BaseMCPClient, MCPClient
from .transport import HttpTransport, WebSocketTransport, MCPTransport
from .types import TransportConfig, ClientConfig, ContextInfo


class MCPClientFactory:
    """Factory for creating MCP v2 clients with different configurations."""

    @staticmethod
    def create_http_client(endpoint: str) -> "MCPClientBuilder":
        """Create an HTTP-based MCP client."""
        transport = TransportConfig(type="http", endpoint=endpoint)
        return MCPClientBuilder(transport)

    @staticmethod
    def create_websocket_client(endpoint: str) -> "MCPClientBuilder":
        """Create a WebSocket-based MCP client."""
        transport = TransportConfig(type="websocket", endpoint=endpoint)
        return MCPClientBuilder(transport)

    @staticmethod
    def create_production_client(transport_type: str, endpoint: str) -> "MCPClientBuilder":
        """Create a production-ready MCP client with optimized settings."""
        transport = TransportConfig(
            type=transport_type,
            endpoint=endpoint,
            timeout=30,
            retry_attempts=3,
            enable_metrics=True,
        )
        return (
            MCPClientBuilder(transport)
            .enable_metrics(True)
            .enable_logging(True)
            .enable_caching(True)
        )

    @staticmethod
    def create_development_client(transport_type: str, endpoint: str) -> "MCPClientBuilder":
        """Create a development MCP client with debug settings."""
        transport = TransportConfig(
            type=transport_type,
            endpoint=endpoint,
            timeout=60,
            retry_attempts=1,
            enable_metrics=True,
        )
        return (
            MCPClientBuilder(transport)
            .enable_metrics(True)
            .enable_logging(True)
            .enable_caching(False)
        )

    @staticmethod
    def create_cached_client(transport_type: str, endpoint: str) -> "MCPClientBuilder":
        """Create a cached MCP client that caches responses."""
        return MCPClientFactory.create_production_client(transport_type, endpoint).enable_caching(
            True
        )


class MCPClientBuilder:
    """Builder for configuring and creating MCP clients."""

    def __init__(self, transport: TransportConfig) -> None:
        self._client_id = str(uuid4())
        self._transport = transport
        self._default_context: Optional[ContextInfo] = None
        self._enable_caching = True
        self._enable_metrics = True
        self._enable_logging = True
        self._metadata: Dict[str, Any] = {}

    def client_id(self, client_id: str) -> "MCPClientBuilder":
        """Set the client ID."""
        self._client_id = client_id
        return self

    def transport(self, transport: TransportConfig) -> "MCPClientBuilder":
        """Set the transport configuration."""
        self._transport = transport
        return self

    def default_context(self, context: ContextInfo) -> "MCPClientBuilder":
        """Set the default context."""
        self._default_context = context
        return self

    def enable_caching(self, enable_caching: bool) -> "MCPClientBuilder":
        """Enable or disable caching."""
        self._enable_caching = enable_caching
        return self

    def enable_metrics(self, enable_metrics: bool) -> "MCPClientBuilder":
        """Enable or disable metrics."""
        self._enable_metrics = enable_metrics
        return self

    def enable_logging(self, enable_logging: bool) -> "MCPClientBuilder":
        """Enable or disable logging."""
        self._enable_logging = enable_logging
        return self

    def metadata(self, key: str, value: Any) -> "MCPClientBuilder":
        """Set client metadata."""
        self._metadata[key] = value
        return self

    def build(self) -> MCPClient:
        """Build the MCP client."""
        config = ClientConfig(
            client_id=self._client_id,
            transport=self._transport,
            default_context=self._default_context,
            enable_caching=self._enable_caching,
            enable_metrics=self._enable_metrics,
            enable_logging=self._enable_logging,
            metadata=self._metadata,
        )

        transport = self._create_transport(self._transport)
        return BaseMCPClient(config, transport)

    def _create_transport(self, config: TransportConfig) -> MCPTransport:
        """Create the appropriate transport based on configuration."""
        transport_type = config.type.lower()
        
        if transport_type == "http":
            return HttpTransport(config)
        elif transport_type == "websocket":
            return WebSocketTransport(config)
        else:
            raise ValueError(f"Unsupported transport type: {config.type}")
